src/lab3/PPMRenderer.ts

```diff
@@ -0,0 +1,69 @@
+import { createWriteStream } from 'fs';
+import { Renderer } from '../lab1/types/Renderer';
+import { Scene } from '../lab1/types/Scene';
+import { WriterPPM } from '../lab2/plugins/ppm/WriterPPM';
+import { ImageBuffer } from '../lab2/ImageBuffer';
+import { PassThrough } from 'stream';
+import Ray from '../lab1/structures/ray/Ray';
+import { Hit } from '../lab1/types/Hit';
+import { findCloserHit } from '../lab1/utils/findCloserHit';
+import { Pixel } from '../lab2/interfaces/Pixel';
+
+export default class PPMRenderer implements Renderer {
+  private writerPPM = new WriterPPM();
+
+  constructor(public readonly scene: Scene, private filePath: string) {}
+
+  public render() {
+    const { camera, objects } = this.scene;
+
+    const writeStream = createWriteStream(this.filePath);
+
+    const pixelsStream = new PassThrough({ objectMode: true });
+
+    const imageBuffer = new ImageBuffer(
+      {
+        height: camera.vResolution,
+        width: camera.hResolution,
+        maxColor: 255,
+      },
+      pixelsStream
+    );
+
+    const PPMStream = this.writerPPM.write(imageBuffer);
+    PPMStream.pipe(writeStream);
+
+    for (let y = 0; y < camera.vResolution; y++) {
+      for (let x = 0; x < camera.hResolution; x++) {
+        const screenPixelPosition = camera.getScreenPixelCoordinates(x, y);
+        const ray = new Ray(
+          camera.focalPoint,
+          screenPixelPosition.toVector().subtract(camera.focalPoint.toVector())
+        );
+        let closestHit: Hit | null = null;
+
+        for (const object of objects) {
+          const hit = object.getIntersection(ray);
+
+          if (!hit) continue;
+
+          closestHit = closestHit ? findCloserHit(hit, closestHit) : hit;
+        }
+
+        const pixel = this.handleHit(closestHit);
+
+        pixelsStream.push(pixel);
```

Comment:
I think most of this ray tracing functionality it common for all Renderes. The only difference is in handleHit. I'd turn the existing CommonRenderer into an abstract class that implements render but doesn't implement renderHit and then extend it to reduce code duplication. You'd also have to put pixelsStream.push(...) into handleHit. And initialize imageBuffer/PPMWriter in the constructor then.

Reply:
Also, you might even more reduce code duplication by creating a CommonImageRenderer class that implements CommonRenderer. You'd just have to pass ImageWriter instance to the constructor and it'd operate on it without even having to know what format this is. To avoid creating unecessary PNGRenderer + BMPRenderer in the future